enhancements/ocp-coreos-layering/ocp-coreos-layering.md

```diff
@@ -0,0 +1,730 @@
+
+For now we're assuming we will not have 100% reproducible images and are planning accordingly.
+
+#### Future Upgrade flow
+1. The "base image" will be part of the release image, the same as today.
+1. The CVO will replace a ConfigMap in the MCO namespace with the OS payload reference, as it does today.
+1. The MCO will update an `imagestream` object (e.g. `openshift-machine-config-operator/rhel-coreos`) when this ConfigMap changes.
+1. The MCO builds a container image for each `MachineConfigPool`, using the new base image
+
+1. Each machineconfig pool will support a `custom-external-coreos` imagestream for pulling externally built images
+1.MCD continues to perform drains and reboots,, but for the most part does not change files in `/etc` or install packages per node
+1. The Machine Configuration Server (MCS) will only serve a "bootstrap" Ignition configuration (pull secret, network configuration) sufficient for the node to pull the target container image.
+
+For clusters without any custom MachineConfig at all, the MCO will deploy the result of the `mco-coreos` build.
+
+<!-- current drawing: https://docs.google.com/drawings/d/1cGly8mjYDezVjUQiUEWd63IO_S2ZrgSbv1cT5X1KJIk/edit -->
+
+#### Preserving `MachineConfig`
+
+We cannot just drop `MachineConfig` as an interface to node configuration. Hence, the MCO will be responsible for starting new builds on upgrades or when new machine config content is rendered.
+
+For most configuration, instead of having the MCD write files on each node, it will be added into the image build run on the cluster.
+To be more specific, most content from the Ignition `systemd/units` and `storage/files` sections (in general, files written into `/etc`) will instead be injected into an internally-generated `Dockerfile` (or equivalent) that performs an effect similar to the example from the [CoreOS layering enhancement](https://github.com/coreos/enhancements/blob/main/os/coreos-layering.md#butane-as-a-declarative-input-format-for-layering).
+
+```dockerfile=
+FROM <coreos>
+# This is needed 
+ADD mco-rendered-config.json /etc/mco-rendered-config.json
+ADD ignition.json /tmp/ignition.json
+RUN ignition-liveapply /tmp/ignition.json && rm -f /tmp/ignition.json
+```
+
+This build process will be tracked via a `mco-coreos-build` `BuildConfig` object which will be monitored by the operator.
+
+The output of this build process will be pushed to the `imagestream/mco-coreos`, which should be used by further build processes.
+
+It should be noted that MachineConfigs will behave exactly the same as they do today. In other words, they will not be subject to the same staging and rollout process that using a custom base image would.
+
+#### Ignition
+
+Ignition will continue to handle the `disks` and `filesystem` sections - for example, LUKS will continue to be applied as it has been today.
+
+However, see below.
```

Comment:
Are you meaning this in the sense that `However, there are some caveats`?

Reply:
right. Colin can correct me if it meant something else.